File: packages/text-recognition/android/src/main/java/io/capawesome/capacitorjs/plugins/mlkit/textrecognition/TextRecognition.java
package io.capawesome.capacitorjs.plugins.mlkit.textrecognition;

import android.net.Uri;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.google.mlkit.vision.common.InputImage;
import com.google.mlkit.vision.text.TextRecognizer;
import com.google.mlkit.vision.text.TextRecognizerOptionsInterface;
import com.google.mlkit.vision.text.chinese.ChineseTextRecognizerOptions;
import com.google.mlkit.vision.text.devanagari.DevanagariTextRecognizerOptions;
import com.google.mlkit.vision.text.japanese.JapaneseTextRecognizerOptions;
import com.google.mlkit.vision.text.korean.KoreanTextRecognizerOptions;
import com.google.mlkit.vision.text.latin.TextRecognizerOptions;
import io.capawesome.capacitorjs.plugins.mlkit.textrecognition.classes.options.ProcessImageOptions;
import io.capawesome.capacitorjs.plugins.mlkit.textrecognition.classes.results.ProcessImageResult;
import io.capawesome.capacitorjs.plugins.mlkit.textrecognition.interfaces.NonEmptyResultCallback;

public class TextRecognition {

    @NonNull
    private final TextRecognitionPlugin plugin;

    public TextRecognition(@NonNull TextRecognitionPlugin plugin) {
        this.plugin = plugin;
    }

    public void processImage(@NonNull ProcessImageOptions options, @NonNull NonEmptyResultCallback<ProcessImageResult> callback) {
        InputImage inputImage = createInputImageFromFilePath(options.getPath());
        if (inputImage == null) {
            callback.error(new Exception(TextRecognitionPlugin.ERROR_LOAD_IMAGE_FAILED));
            return;
        }

        TextRecognizerOptionsInterface recognizerOptions = createRecognizerOptionsForScript(options.getScript());
        final TextRecognizer textRecognizer = com.google.mlkit.vision.text.TextRecognition.getClient(recognizerOptions);
        plugin
            .getActivity()
            .runOnUiThread(() -> {
                textRecognizer
                    .process(inputImage)
                    .addOnSuccessListener(text -> {
                        textRecognizer.close();
                        ProcessImageResult result = new ProcessImageResult(text);
                        callback.success(result);
                    })
                    .addOnCanceledListener(() -> {
                        textRecognizer.close();
                        callback.error(new Exception(TextRecognitionPlugin.ERROR_PROCESS_IMAGE_CANCELED));
                    })
                    .addOnFailureListener(exception -> {
                        textRecognizer.close();
                        callback.error(exception);
                    });
            });
    }

    @NonNull
    private TextRecognizerOptionsInterface createRecognizerOptionsForScript(@NonNull String script) {
        switch (script) {
            case "CHINESE":
                return new ChineseTextRecognizerOptions.Builder().build();
            case "DEVANAGARI":
                return new DevanagariTextRecognizerOptions.Builder().build();
            case "JAPANESE":
                return new JapaneseTextRecognizerOptions.Builder().build();
            case "KOREAN":
                return new KoreanTextRecognizerOptions.Builder().build();
            default:
                return TextRecognizerOptions.DEFAULT_OPTIONS;
        }
    }

    @Nullable
    private InputImage createInputImageFromFilePath(@NonNull String path) {
        try {
            return InputImage.fromFilePath(plugin.getContext(), Uri.parse(path));
        } catch (Exception exception) {
            return null;
        }
    }
}

File: packages/text-recognition/ios/Plugin/TextRecognition.swift
import Foundation
import MLKitVision
import MLKitTextRecognition
import MLKitTextRecognitionChinese
import MLKitTextRecognitionDevanagari
import MLKitTextRecognitionJapanese
import MLKitTextRecognitionKorean

@objc public class TextRecognition: NSObject {
    public let plugin: TextRecognitionPlugin

    init(plugin: TextRecognitionPlugin) {
        self.plugin = plugin
    }

    @objc func processImage(_ options: ProcessImageOptions, completion: @escaping (ProcessImageResult?, Error?) -> Void) {
        guard let visionImage = createVisionImageFromFilePath(options.getPath()) else {
            completion(nil, CustomError.loadImageFailed)
            return
        }

        let recognizerOptions = createRecognizerOptionsForScript(options.getScript())
        let textRecognizer = TextRecognizer.textRecognizer(options: recognizerOptions)
        textRecognizer.process(visionImage) { text, error in
            if let error = error {
                completion(nil, error)
                return
            }
            guard let text = text else {
                completion(nil, CustomError.loadImageFailed)
                return
            }
            let result = ProcessImageResult(text: text)
            completion(result, nil)
        }
    }

    private func createRecognizerOptionsForScript(_ script: String) -> CommonTextRecognizerOptions {
        switch script {
        case "CHINESE":
            return ChineseTextRecognizerOptions()
        case "DEVANAGARI":
            return DevanagariTextRecognizerOptions()
        case "JAPANESE":
            return JapaneseTextRecognizerOptions()
        case "KOREAN":
            return KoreanTextRecognizerOptions()
        default:
            return TextRecognizerOptions()
        }
    }

    private func createVisionImageFromFilePath(_ path: String) -> VisionImage? {
        guard let url = URL.init(string: path) else {
            return nil
        }
        if FileManager.default.fileExists(atPath: url.path) {
            guard let image = UIImage.init(contentsOfFile: url.path) else {
                return nil
            }
            let visionImage = VisionImage.init(image: image)
            visionImage.orientation = image.imageOrientation
            return visionImage
        } else {
            return nil
        }
    }
}
